### PR TITLE
HHH-10742

### PR DESCRIPTION
**Pull request by brmeyer**

### CODE REVIEW

File: hibernate-core/src/test/java/org/hibernate/test/batchfetch/Model.java [hunk 1]
```diff
@@ -18,7 +18,7 @@
 	private String description;
 	private ProductLine productLine;
 	
-	Model() {}
+	public Model() {}
```
Comment (brmeyer) [inline hibernate-core/src/test/java/org/hibernate/test/batchfetch/Model.java]: @sebersole, this is my main point of confusion.  The non-public ctor caused the following Exception with the new CL setup in `JavassistProxyFactory`.  There are several other failures in hibernate-core for similar reasons.

`Caused by: java.lang.IllegalAccessError: tried to access method org.hibernate.test.batchfetch.Model.<init>()V from class org.hibernate.test.batchfetch.Model_$$_jvst960_2
	at org.hibernate.test.batchfetch.Model_$$_jvst960_2.<init>(Model_$$_jvst960_2.java)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:422)
	at java.lang.Class.newInstance(Class.java:442)
	at org.hibernate.proxy.pojo.javassist.JavassistProxyFactory.getProxy(JavassistProxyFactory.java:136)`

Super confused why that only came up now, and not before...

**Comment by sebersole** [comment]
That is really strange considering that the generated proxy class (Model_$$jvst960_2) is a subclass of the class in question (org.hibernate.test.batchfetch.Model) *and* they are defined in the same package.

**Comment by brmeyer** [comment]
@sebersole Probably due to to the persistentClass and the proxy using two different ClassLoaders, right?  See Javassist's `ProxyFactory#getClassLoader0`.  This is checking the superclass (ie, the persistentClass), all interfaces, etc.  Originally, we were sidestepping the whole thing and always returning the persistentClass' ClassLoader.  Now, we're returning an anonymous ClassLoader and may be causing the conflict?

I've been experimenting with different configurations and am failing to find something that works for all cases.  Admittedly, the internals of proxying, Javassist, etc. makes my head spin, so maybe I'm missing something obvious...

@scottmarlow, any thoughts as well?

**Comment by scottmarlow** [comment]
@brmeyer in terms of hacking the current ORM code up, maybe a different set of classes that are used for OSGi, so that you are only changing the classloader for OSGi.  I'm not sure if all modular (jigsaw, wildfly, osgi) could share the same classloader solution.  Either way, I think that a requirement is that only one classloader is allowed to load any of the application classes, otherwise, we likely would see ClassCastExceptions if multiple class definitions for an application class, is referenced.  

Regarding the "java.lang.IllegalAccessError: tried to access method" error, where is that thrown from?

**Comment by sebersole** [comment]
@brmeyer does this still need merged?  And is is ready to merge?

@scottmarlow Only one accessible ClassLoader should physically load a class.  But that is actually the principal that Hibernate's ClassLoaderService works under.  It simply delegates requests to load a class by name to the individual ClassLoaders it has aggregated together.  Yes, this can still cause a problem, but only in the scenario where:

1. "Foo" has already been loaded by ClassLoader CL1
2. We are asked to find class "Foo" (AggregatedClassLoader#findClass override); as we iterate the aggregated ClassLoaders we ask the first ClassLoader (CL2) to load "Foo" (ClassLoader#loadClass) and it can.

Now we have 2 distinct Class<Foo> references - distinct in that they were each loaded from different ClassLoaders.

We could get around that maybe by first asking each aggregated ClassLoader to find the Class, and then doing a load.  At least in my less-than-crystal-clear understanding of class loading :)

**Comment by brmeyer** [comment]
Hey @sebersole, definitely not ready to merge.  That public ctor fix corrected only one of several failures.  Wanted to get to the bottom of the issue, but I'm still fairly confused about how the ClassLoader we're providing `ProxyFactory` may be causing it...

**Comment by beikov** [comment]
Hey @brmeyer 

if this is still an issue, please reopen a new PR against master.

**Pull request closed without merging**